prov/gni/src/gnix_buddy_allocator.c

```diff
@@ -0,0 +1,353 @@
+				  alloc_handle->lists +
+				  alloc_handle->nlists - 1);
+		offset += alloc_handle->max;
+	}
+
+	return FI_SUCCESS;
+}
+
+/**
+ * Split a block in list "j" until list "i" is reached.
+ */
+static inline void __gnix_buddy_split(handle_t *alloc_handle, size_t j, size_t i)
+{
+	void *tmp = NULL;
+
+	dlist_remove(tmp = alloc_handle->lists[j].next);
+
+	_gnix_set_bit(&alloc_handle->bitmap,
+		      BITMAP_INDEX(tmp, alloc_handle->base, MIN_BLOCK_SIZE,
+				   OFFSET(MIN_BLOCK_SIZE, j)));
+
+
+	/* Split the block until we reach list "i" */
+	for (j--; j > i; j--) {
+		_gnix_set_bit(&alloc_handle->bitmap,
+			      BITMAP_INDEX(tmp +
+					   OFFSET(MIN_BLOCK_SIZE, j),
+					   alloc_handle->base,
+					   MIN_BLOCK_SIZE,
+					   OFFSET(MIN_BLOCK_SIZE, j)));
+
+		dlist_insert_tail(tmp + OFFSET(MIN_BLOCK_SIZE, j),
+				  alloc_handle->lists + j);
+	}
+
+	/* Insert last block into list "i" */
+	dlist_insert_head(tmp, alloc_handle->lists + j);
+
+	/* Insert the buddy block of tmp into list "i" */
+	dlist_insert_tail(tmp + OFFSET(MIN_BLOCK_SIZE, j),
+			  alloc_handle->lists + j);
+}
+
+/**
+ * Find the first free block that can be split, then split it.
+ *
+ * @return 1  if the block cannot be found.
+ *
+ * @return 0 if the block is found.
+ */
+static inline int __gnix_buddy_find_block(handle_t *alloc_handle, size_t i)
+{
+	size_t j;
+
+	for (j = i + 1; j < alloc_handle->nlists; j++) {
+		if (!dlist_empty(alloc_handle->lists + j)) {
+			__gnix_buddy_split(alloc_handle, j, i);
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+
+/**
+ * If the buddy block is on the free list then coalesce and insert into the next
+ * list until we reach an allocated or split buddy block, or the max list size.
+ */
+static inline void __gnix_buddy_coalesce(handle_t *alloc_handle, void **ptr,
+					 size_t *block_size)
+{
+	while (*block_size < alloc_handle->max &&
+	       !_gnix_test_bit(&alloc_handle->bitmap,
+			       BITMAP_INDEX(BUDDY(*ptr, *block_size,
+						  alloc_handle->base),
+					    alloc_handle->base,
+					    MIN_BLOCK_SIZE,
+					    *block_size))) {
+
+		dlist_remove(BUDDY(*ptr, *block_size, alloc_handle->base));
+
+		/* Ensure ptr is the beginning of the new block */
+		if (*ptr > BUDDY(*ptr, *block_size, alloc_handle->base)) {
+			*ptr = BUDDY(*ptr, *block_size, alloc_handle->base);
+		}
+
+		*block_size *= 2;
+
+		_gnix_clear_bit(&alloc_handle->bitmap,
+				BITMAP_INDEX(*ptr, alloc_handle->base,
+					     MIN_BLOCK_SIZE, *block_size));
+	}
+
+}
+
+int _gnix_buddy_allocator_create(void *base, size_t len, size_t max,
+				 handle_t **alloc_handle)
+{
+	char err_buf[256] = {0}, *error = NULL;
+	int fi_errno;
+
+	GNIX_TRACE(FI_LOG_EP_CTRL, "\n");
+
+	/* Ensure parameters are valid */
+	if (!base || !len || !max || max > len || !alloc_handle ||
+	    IS_NOT_POW_TWO(max) || (len % max) || !(len / MIN_BLOCK_SIZE * 2)) {
+
+		GNIX_WARN(FI_LOG_EP_CTRL,
+			  "Invalid parameter to _gnix_buddy_allocator_create."
+			  "\n");
```

Comment:
Why do you add newlines to some of the print statements (and then why not to others)?